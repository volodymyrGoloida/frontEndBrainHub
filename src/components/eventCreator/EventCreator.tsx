import React, { useState } from "react";
import { createEventDate } from "../../cms/cms";
import { EventCreateRequest } from "../../cms/types";
import { validateEmptyFields } from "../../utils/checkEmptiesFields";
import { isValidEmail, isValidEmailLength } from "../../utils/validateEmail";
import { CustomInput, Button } from "../UIComponents";
import scss from "./EventCreator.module.scss";

const EventCreator: React.FC = () => {
  const [eventCreate, setEventCreate] = useState<EventCreateRequest>({
    firstName: "",
    lastName: "",
    email: "",
    eventDate: "",
  });
  const [eventCreateErrors, setEventCreateErrors] =
    useState<EventCreateRequest>({
      firstName: "",
      lastName: "",
      email: "",
      eventDate: "",
    });

  const handleEventCreate = (name: string, value: string) => {
    setEventCreateErrors((prevValue) => ({
      ...prevValue,
      [name]: "",
    }));
    setEventCreate((prevValues) => ({
      ...prevValues,
      [name]: value,
    }));
  };

  const handleCreateEvent = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const eventDataEmptyFields =
      validateEmptyFields<EventCreateRequest>(eventCreate);
    if (Object.keys(eventDataEmptyFields).length > 0) {
      setEventCreateErrors(eventDataEmptyFields);
      return;
    }

    if (!isValidEmailLength(eventCreate.email)) {
      setEventCreateErrors((prevValue) => ({
        ...prevValue,
        email: "Length for email can be at least 3 elements.",
      }));
      return;
    }

    if (!isValidEmail(eventCreate.email)) {
      setEventCreateErrors((prevValue) => ({
        ...prevValue,
        email: "Your email isn't correct, please check it.",
      }));
      return;
    }

    const statusCode = await createEventDate({
      ...eventCreate,
      eventDate: new Date(eventCreate.eventDate).getTime().toString(),
    });

    if (statusCode && statusCode === 200) {
      alert("Your event was created");
    }
    setEventCreate({
      firstName: "",
      lastName: "",
      email: "",
      eventDate: "",
    });
  };

  return (
    <div className={scss.formWrapper}>
      <form onSubmit={handleCreateEvent}>
        <h1>Create your own event!</h1>
        <div className={scss.inputWrapper}>
          <CustomInput
            onChange={handleEventCreate}
            error={eventCreateErrors.firstName}
            name="firstName"
            value={eventCreate.firstName}
            type="text"
            placeholder="please enter your first name"
          />
        </div>
        <div className={scss.inputWrapper}>
          <CustomInput
            onChange={handleEventCreate}
            error={eventCreateErrors.lastName}
            name="lastName"
            value={eventCreate.lastName}
            type="text"
            placeholder="please enter your last name"
          />
        </div>
        <div className={scss.inputWrapper}>
          <CustomInput
            onChange={handleEventCreate}
            name="email"
            value={eventCreate.email}
            error={eventCreateErrors.email}
            type="text"
            placeholder="please enter your email"
          />
        </div>
        <div className={scss.inputWrapper}>
          <CustomInput
            value={eventCreate.eventDate}
            onChange={handleEventCreate}
            error={eventCreateErrors.eventDate}
            name="eventDate"
            type="datetime-local"
          />
        </div>
        <div className={scss.buttonWrapper}>
          <Button>Create</Button>
        </div>
      </form>
    </div>
  );
};
export default EventCreator;
