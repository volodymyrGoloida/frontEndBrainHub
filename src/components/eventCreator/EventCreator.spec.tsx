import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import EventCreator from "./EventCreator";

jest.mock("../../cms/cms");

describe("component EventCreator", () => {
  it("should correct went through all user flow", async () => {
    const { baseElement } = render(<EventCreator />);
    expect(baseElement).toBeDefined();

    const buttonEl = screen.getByText("Create");
    expect(buttonEl).toBeInTheDocument();

    fireEvent.click(buttonEl);

    let inputErrors = screen.getAllByText("This field is empty");
    inputErrors.forEach((el) => expect(el).toBeInTheDocument());
    expect(inputErrors.length).toBe(4);

    const firstNameInput = screen.getByPlaceholderText(
      "please enter your first name"
    ) as HTMLInputElement;
    fireEvent.change(firstNameInput, {
      target: {
        value: "q",
      },
    });
    inputErrors = screen.getAllByText("This field is empty");
    expect(inputErrors.length).toBe(3);

    const emailInput = screen.getByPlaceholderText(
      "please enter your email"
    ) as HTMLInputElement;
    const lastNameInput = screen.getByPlaceholderText(
      "please enter your last name"
    ) as HTMLInputElement;

    // eslint-disable-next-line testing-library/no-node-access
    const datetimeLocalInput = baseElement.querySelector(
      "[name='eventDate']"
    ) as HTMLInputElement;

    fireEvent.change(lastNameInput, {
      target: {
        value: "qw",
      },
    });
    fireEvent.change(emailInput, {
      target: {
        value: "qw",
      },
    });
    fireEvent.change(datetimeLocalInput, {
      target: {
        value: "2018-06-12T19:30",
      },
    });

    fireEvent.click(buttonEl);
    let emailInputError = screen.getByText(
      "Length for email can be at least 3 elements."
    );
    expect(emailInputError).toBeInTheDocument();

    fireEvent.change(emailInput, {
      target: {
        value: "qwe3;@com",
      },
    });
    fireEvent.click(buttonEl);
    emailInputError = screen.getByText(
      "Your email isn't correct, please check it."
    );
    expect(emailInputError).toBeInTheDocument();

    fireEvent.change(emailInput, {
      target: {
        value: "qweqwe123@gmail.com",
      },
    });
    fireEvent.click(buttonEl);

    await waitFor(() => {
      expect(emailInput.value).toBe("");
    });
    expect(firstNameInput.value).toBe("");
    expect(lastNameInput.value).toBe("");
    expect(datetimeLocalInput.value).toBe("");
    expect(baseElement).toMatchSnapshot();
  });
});
