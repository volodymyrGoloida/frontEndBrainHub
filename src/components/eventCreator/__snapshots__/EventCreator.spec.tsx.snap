// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component EventCreator should correct went through all user flow 1`] = `
<body>
  <div>
    <div
      class="formWrapper"
    >
      <form>
        <h1>
          Create your own event!
        </h1>
        <div
          class="inputWrapper"
        >
          <div
            class="root"
          >
            <input
              autocomplete="off"
              name="firstName"
              placeholder="please enter your first name"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="inputWrapper"
        >
          <div
            class="root"
          >
            <input
              autocomplete="off"
              name="lastName"
              placeholder="please enter your last name"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="inputWrapper"
        >
          <div
            class="root"
          >
            <input
              autocomplete="off"
              name="email"
              placeholder="please enter your email"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="inputWrapper"
        >
          <div
            class="root"
          >
            <input
              autocomplete="off"
              name="eventDate"
              type="datetime-local"
              value=""
            />
          </div>
        </div>
        <div
          class="buttonWrapper"
        >
          <button
            class="buttonRoot"
          >
            Create
          </button>
        </div>
      </form>
    </div>
  </div>
</body>
`;
